src/allmydata/test/test_storage_https.py

```diff
@@ -0,0 +1,232 @@
+            # is necessary when it's not for IStorageServer HTTPS tests.
+            await deferLater(reactor, 0.001)
+
+    def request(self, url: str, expected_certificate: x509.Certificate):
+        """
+        Send a HTTPS request to the given URL, ensuring that the given
+        certificate is the one used via SPKI-hash-based pinning comparison.
+        """
+        # No persistent connections, so we don't have dirty reactor at the end
+        # of the test.
+        treq_client = HTTPClient(
+            Agent(
+                reactor,
+                _StorageClientHTTPSPolicy(
+                    expected_spki_hash=get_spki_hash(expected_certificate)
+                ),
+                pool=HTTPConnectionPool(reactor, persistent=False),
+            )
+        )
+        return treq_client.get(url)
+
+    @async_to_deferred
+    async def test_success(self):
+        """
+        If all conditions are met, a TLS client using the Tahoe-LAFS policy can
+        connect to the server.
+        """
+        private_key = self.generate_private_key()
+        certificate = self.generate_certificate(private_key)
+        async with self.listen(
+            self.to_file(private_key), self.to_file(certificate)
+        ) as url:
+            response = await self.request(url, certificate)
+            self.assertEqual(await response.content(), b"YOYODYNE")
+
+    @async_to_deferred
+    async def test_server_certificate_has_wrong_hash(self):
+        """
+        If the server's certificate hash doesn't match the hash the client
+        expects, the request to the server fails.
+        """
+        private_key1 = self.generate_private_key()
+        certificate1 = self.generate_certificate(private_key1)
+        private_key2 = self.generate_private_key()
+        certificate2 = self.generate_certificate(private_key2)
+
+        async with self.listen(
+            self.to_file(private_key1), self.to_file(certificate1)
+        ) as url:
+            with self.assertRaises(ResponseNeverReceived):
+                await self.request(url, certificate2)
+
+    @async_to_deferred
+    async def test_server_certificate_expired(self):
+        """
+        If the server's certificate has expired, the request to the server
+        succeeds if the hash matches the one the client expects; expiration has
+        no effect.
+        """
+        private_key = self.generate_private_key()
+        certificate = self.generate_certificate(private_key, expires_days=-10)
+
+        async with self.listen(
+            self.to_file(private_key), self.to_file(certificate)
+        ) as url:
+            response = await self.request(url, certificate)
+            self.assertEqual(await response.content(), b"YOYODYNE")
+
+    # TODO an obvious attack is a private key that doesn't match the
+    # certificate... but OpenSSL (quite rightly) won't let you listen with that
+    # so I don't know how to test that!
```

Comment:
Hm.  So, the way the TLS handshake goes is ... (a bunch of stuff) ... then the client sends a "ClientKeyExchange" message which has a "PreMasterSecret" in it, encrypted using the public key from the server's certificate.  The server must decrypt this and do a cipher-specific operation to derive the "MasterSecret".  Later the server sends its "Finished" message encrypted using "MasterSecret".  The client decrypts it to extract a hash and a MAC of all of the handshake messages up until this point.

Without the private key corresponding to the certificate it presents, the server must compute the wrong "MasterSecret" and when the client decrypts the "Finished" message with the correct "MasterSecret" the hash and MAC will mismatch and the handshake will fail.

So ... a test for "the server is using the wrong private key" seems like it would mostly be a test for the underlying TLS implementation (which, at worst, would be able to deliver garbage data to the client instead of properly signaling a handshake failure).

However, I probably wouldn't mind seeing a test for this case if it were possible to write one since perhaps some TLS libraries offer some ways to choose ciphers so poorly that you somehow lose the above protection (I *think* this would have to be roughly a choice of the NULL cipher everywhere which would be a pretty bad mistake, but ...).

That said, I don't really know how we'd write this test, given OpenSSL's behavior.  My only idea is ... find a different TLS implementation that lets you do something so obviously illegal?

It's probably fine to defer any such work to a follow-up ticket though.

Reply:
> perhaps some TLS libraries offer some ways to choose ciphers so poorly that you somehow lose the above protection

This logic, as a reminder, is client side. And the test would be of the client's behavior. So I'm not sure this test adds very much over .. auditing we've configured the correct ciphers?